python/lsst/daf/butler_migrate/script/update_day_obs.py

```diff
@@ -0,0 +1,165 @@
+
+    Returns
+    -------
+    new : `lsst.daf.butler.DimensionRecord`
+       New record with updated values.
+    """
+    d = rec.toDict()
+    # Should validate but this is not a general function for now.
+    d.update(kwargs)
+    return type(rec)(**d)
+
+
+def update_day_obs(repo: str, instrument: str) -> None:
+    """Update the day_obs for the given instrument.
+
+    Parameters
+    ----------
+    repo : `str`
+        URI of butler repository to update.
+    instrument : `str`
+        Name of instrument to use to update records.
+
+    Notes
+    -----
+    This update code depends on being able to import the python instrument
+    class registered with the instrument record.
+    """
+    # Connect to the butler.
+    butler = Butler.from_config(repo, writeable=True)
+
+    # Need the instrument class, since that is used to calculate day_obs.
+    # Do not depend directly on pipe_base but to load this class pipe_base
+    # will have to be available.
+    instr_records = list(butler.registry.queryDimensionRecords("instrument", instrument=instrument))
+    if len(instr_records) != 1:
+        if not instr_records:
+            raise RuntimeError(f"Unable to find an instrument record for instrument named {instrument}.")
+        else:
+            raise AssertionError(f"Impossibly got more than one instrument record named {instrument}.")
+    instrument_class_name = instr_records[0].class_name
+    instrument_cls = doImportType(instrument_class_name)
+    instr = instrument_cls()
+    translator = instr.translatorClass
+    if translator is None:
+        raise RuntimeError(
+            f"Instrument class {instrument_class_name} has no registered translator class."
+            " Unable to calculate the correct observing day."
+        )
+
+    # The naive approach is to query all the records and recalculate the
+    # day_obs and then update them all in one big transaction. This will
+    # be fine for a small repo but catastrophic with millions of exposure
+    # and visit records.
+    exposures_to_be_updated = {}
+    exposures = butler.registry.queryDimensionRecords("exposure", instrument=instrument)
+    counter = 0
+    for exp in exposures:
+        offset = translator.observing_date_to_offset(exp.timespan.begin)
+        day_obs = translator.observing_date_to_observing_day(exp.timespan.begin, offset)
+        if day_obs != exp.day_obs:
+            # Need to update the record. Immutable so need a copy.
+            exposures_to_be_updated[exp.id] = updated_record(exp, day_obs=day_obs)
+        counter += 1
+
+    _LOG.info(
+        "Number of exposure records needing to be updated: %d / %d", len(exposures_to_be_updated), counter
+    )
+
+    # If there are modified exposures associated visits will have to be
+    # located and updated.
+    visit_defs = butler.registry.queryDimensionRecords(
+        "visit_definition",
+        where="exposure in (exps)",
+        bind={"exps": list(exposures_to_be_updated)},
```

Comment:
This potentially needs to be in batches too, but if you are sure that it's not going to be large then it's OK.

Reply:
I've been assuming that `queryDimensionRecords` would be clever enough to do the batching internally if the list is too long, but if you think that's going to be problematic I can chunk it externally.

Reply:
My concern here is that `WHERE exposure in (...)` will become very long, and we do not batch that expression internally.